llrt_core/src/modules/http/url.rs

```diff
@@ -2,400 +2,260 @@ use std::{path::PathBuf, str::FromStr};
-                Ok(s) => s,
-                Err(_) => return Null.into_js(&ctx),
-            };
-            let path_string = match get_string(&ctx, input) {
-                Ok(s) => s,
-                Err(_) => return Null.into_js(&ctx),
-            };
-
-            match base_string.parse::<Url>() {
-                Ok(base_url) => {
-                    if let Ok(parsed_url) = base_url.join(&path_string) {
-                        return URL::create(ctx.clone(), parsed_url).unwrap().into_js(&ctx);
-                    }
-                },
-                Err(_) => return Null.into_js(&ctx),
-            }
-        } else if input.is_string() {
-            if let Ok(string_val) = input.get::<String>() {
-                if let Ok(parsed_url) = Url::parse(&string_val) {
-                    return URL::create(ctx.clone(), parsed_url).unwrap().into_js(&ctx);
-                }
-            }
-        }
-        Null.into_js(&ctx)
+    #[qjs(get)]
+    pub fn hash(&self) -> String {
+        quirks::hash(&self.url.borrow()).to_string()
     }
 
-    pub fn to_string(&self) -> String {
-        self.format(true, true, true, false)
+    #[qjs(set, rename = "hash")]
+    pub fn set_hash(&mut self, hash: String) -> String {
+        quirks::set_hash(&mut self.url.borrow_mut(), hash.as_str());
+        hash
     }
 
     #[qjs(get)]
-    fn search_params(&self) -> Class<'js, URLSearchParams> {
-        self.search_params.clone()
+    pub fn host(&self) -> String {
+        quirks::host(&self.url.borrow()).to_string()
     }
 
-    #[qjs(get)]
-    fn href(&self) -> String {
-        self.to_string()
+    #[qjs(set, rename = "host")]
+    pub fn set_host(&mut self, host: Coerced<String>) -> String {
+        let _ = quirks::set_host(&mut self.url.borrow_mut(), host.as_str());
+        host.0
     }
 
-    #[qjs(set, rename = "href")]
-    fn set_href(&mut self, ctx: Ctx<'js>, href: String) -> Result<String> {
-        let new = Self::from_str(ctx, &href)?;
-
-        self.protocol = new.protocol;
-        self.host = new.host;
-        self.hostname = new.hostname;
-        self.port = new.port;
-        self.pathname = new.pathname;
-        self.hash = new.hash;
-        self.search_params = new.search_params;
-        self.username = new.username;
-        self.password = new.password;
-        Ok(href)
+    #[qjs(get)]
+    pub fn hostname(&self) -> String {
+        quirks::hostname(&self.url.borrow()).to_string()
     }
 
-    #[qjs(get)]
-    fn origin(&self) -> String {
-        format!("{}://{}", &self.protocol, &self.host)
+    #[qjs(set, rename = "hostname")]
+    pub fn set_hostname(&mut self, hostname: Coerced<String>) -> String {
+        // TODO: This should be fixed in Url
+        if HOSTNAME_DELIMITER_REGEX.find(hostname.as_str()).is_none() {
+            let _ = quirks::set_hostname(&mut self.url.borrow_mut(), hostname.as_str());
+        }
+        hostname.0
     }
 
     #[qjs(get)]
-    fn protocol(&self) -> String {
```

Comment:
Wow, how did I miss that? That would be great and I need to go update my own project as well.

Reply:
Yeah, and we need to apply this throughout the project (in another PR obviously)

Reply:
So...I get borrowing errors when I try to switch it:

```rust
    #[qjs(get)]
    pub fn protocol(&self) -> &str {
        quirks::protocol(&self.url.borrow())
    }
```

```
error[E0515]: cannot return value referencing temporary value
   --> llrt_core/src/modules/http/url.rs:187:9
    |
187 |         quirks::protocol(&self.url.borrow())
    |         ^^^^^^^^^^^^^^^^^^-----------------^
    |         |                 |
    |         |                 temporary value created here
    |         returns a value referencing data owned by the current function
```

It seems like because we're borrowing and quirks returns a reference to the borrowed value, we never have ownership and can't return a reference to it. I'm fairly new to Rust -- is there a way to handle it so we're not copying the data?

Reply:
Ah right, I forgot that the value is now owned by the quirks reference. I don't think we can do anything else than copy in this case

Reply:
Bummer. Still great to know that's an option though!